src/blocks.zig

```diff
@@ -414,39 +418,111 @@ pub const Block = packed struct { // MARK: Block
 };
 
 // MARK: Tick
-pub var tickFunctions: utils.NamedCallbacks(TickFunctions, TickFunction) = undefined;
-pub const TickFunction = fn(block: Block, _chunk: *chunk.ServerChunk, x: i32, y: i32, z: i32) void;
-pub const TickFunctions = struct {
-	pub fn replaceWithCobble(block: Block, _chunk: *chunk.ServerChunk, x: i32, y: i32, z: i32) void {
-		std.log.debug("Replace with cobblestone at ({d},{d},{d})", .{x, y, z});
-		const cobblestone = parseBlock("cubyz:cobblestone");
+pub fn TickEventTableMap(comptime Child: type) type {
+	return struct {
+		const Self = @This();
 
-		const wx = _chunk.super.pos.wx + x;
-		const wy = _chunk.super.pos.wy + y;
-		const wz = _chunk.super.pos.wz + z;
+		tableMap: std.StringHashMap(*const TickEvent.VTable) = undefined,
 
-		_ = main.server.world.?.cmpxchgBlock(wx, wy, wz, block, cobblestone);
-	}
+		pub fn init() Self {
+			var self = Self{
+				.tableMap = .init(main.globalAllocator.allocator),
+			};
+
+			switch(@typeInfo(Child)) {
+				.@"struct" => |childStruct| {
+					inline for(childStruct.decls) |childDecl| {
+						const childField = @field(Child, childDecl.name);
+						if(@typeInfo(@TypeOf(childField)) == .type) {
+							registerEventStruct(childField, &self);
+						}
+					}
+				},
+				else => {
+					std.log.err("Type {s} is not a struct", .{@typeName(Child)});
+				},
+			}
+
+			return self;
+		}
+
+		fn registerEventStruct(comptime EventType: type, self: *Self) void {
+			inline for(@typeInfo(EventType).@"struct".decls) |eventDecl| {
+				const field = @field(EventType, eventDecl.name);
+				if(@TypeOf(field) == *const TickEvent.VTable) {
+					const id = @field(EventType, "id");
+					self.tableMap.putNoClobber(id, field) catch unreachable;
+					std.log.debug("Registered Callback '{s}'", .{eventDecl.name});
+				}
+			}
+		}
+
+		pub fn deinit(self: *Self) void {
+			self.tableMap.deinit();
+		}
+
+		pub fn getTablePointer(self: *Self, id: []const u8) ?*const TickEvent.VTable {
+			return self.tableMap.get(id);
+		}
+	};
+}
+
+pub var tickEvents: TickEventTableMap(TickEvents) = undefined;
+pub const TickFunction = fn(self: *anyopaque, block: Block, _chunk: *chunk.ServerChunk, x: i32, y: i32, z: i32) void;
+pub const TickEvents = struct {
+	pub const ReplaceWithEvent = struct {
+		pub const id = "replaceWith";
+		replacementBlock: Block,
+
+		pub const VTable: *const TickEvent.VTable = &.{
+			.id = id,
+			.tickFunction = utils.castFunctionSelfToAnyopaque(replaceWith),
+			.loadFromZon = utils.castFunctionReturnToAnyopaque(loadFromZon),
+		};
+
+		pub fn replaceWith(self: *ReplaceWithEvent, block: Block, _chunk: *chunk.ServerChunk, x: i32, y: i32, z: i32) void {
+			const wx = _chunk.super.pos.wx + x;
+			const wy = _chunk.super.pos.wy + y;
+			const wz = _chunk.super.pos.wz + z;
```

Comment:
Instead of passing the chunk here and then calculating the positions, I think you should just pass wx/wy/wz directly and not pass the chunk at all.